compiler/rustc_codegen_llvm/src/asm.rs

```diff
@@ -125,15 +125,39 @@ impl AsmBuilderMethods<'tcx> for Builder<'a, 'll, 'tcx> {
 
         // Collect the types of output operands
         let mut constraints = vec![];
+        let mut clobbers = vec![];
         let mut output_types = vec![];
         let mut op_idx = FxHashMap::default();
         for (idx, op) in operands.iter().enumerate() {
             match *op {
                 InlineAsmOperandRef::Out { reg, late, place } => {
+                    let is_target_supported = |reg_class: InlineAsmRegClass| {
+                        for &(_, feature) in reg_class.supported_types(asm_arch) {
+                            if let Some(feature) = feature {
+                                if self.tcx.sess.target_features.contains(&Symbol::intern(feature))
+                                {
+                                    return true;
+                                }
+                            } else {
+                                // Register class is unconditionally supported
+                                return true;
+                            }
+                        }
+                        false
+                    };
+
                     let mut layout = None;
                     let ty = if let Some(ref place) = place {
                         layout = Some(&place.layout);
                         llvm_fixup_output_type(self.cx, reg.reg_class(), &place.layout)
+                    } else if !is_target_supported(reg.reg_class()) {
```

Comment:
Why does this need to happen conditionally? Can't we always "just" clobber the clobber-like constraints, regardless of the target features?

Reply:
Just clobbering loses the distinction between `out` and `lateout`: LLVM clobbers act like an `out`, which prevents the registers from being used as an input. However in the case of `lateout` we want to allow the register allocator to use the register being clobbered as an input.

For register classes that are unavailable due to disabled target features this is not a problem: the register can't be used as an input anyways.